### ISSUE TITLE
RFE: add an option to park the rpmdb

### ISSUE THREAD

**Issue by cgwalters** [rpm-software-management/rpm#2219]
This is a followup to https://github.com/rpm-software-management/rpm/commit/71456f2fc09900a027a33dc3d6d75c69a9b39488
which is about generating bit-for-bit reproducible images (container/disk) that include an RPM database.

At the time, the person working on that PR was looking at RHEL8 (BDB format rpm database).  We've since moved on to sqlite, which is nicer.

Except in my testing, this `rpmdb.sqlite-shm` is the very last thing blocking reproducible rpm-ostree container images.  I hacked up https://github.com/coreos/rpm-ostree/pull/4074
but this issue is to track fixing this more properly inside librpm.

I think what would be the best is if this file could be deleted entirely at the end of a build.  It seems like rpm (or sqlite) will auto-recreate it in that case, but what blocks us from doing that is that things fail if the mount is read-only.

**Comment by pmatilai** [comment]
The existence of .sqlite-shm is required for read-only WAL mode to work at all (a very important use-case being queries by regular users), see https://www.sqlite.org/wal.html#read_only_database

**Comment by pmatilai** [comment]
Hmm. Actually doing the WAL docs suggested `PRAGMA journal_mode=DELETE` on the database results in a database that can still be opened by rpm, regular users and all. Rpm will try to re-establish WAL if opened in read-write mode, so nothing too terrible should happen if you do the journal delete as a part of the image building. Except voiding the warranty :wink: 

It'd be nicer of course if rpm had a supported procedure to "park" databases for this kind of thing. --rebuilddb with some special flag maybe.

**Comment by champtar** [comment]
rpmdb reproducibility is also an issue when building (EL9) containers, it would be nice to have a park option indeed.

**Comment by DemiMarie** [comment]
> It'd be nicer of course if rpm had a supported procedure to "park" databases for this kind of thing. --rebuilddb with some special flag maybe.

`--rebuilddb` is much heavier than just a single SQL command.  Perhaps `--parkdb`, along with a corresponding C API function?

**Comment by cgwalters** [comment]
> The existence of .sqlite-shm is required for read-only WAL mode to work at all (a very important use-case being queries by regular users), see https://www.sqlite.org/wal.html#read_only_database

I find this weird - because unprivileged code can't write directly to the database, what practical use is the shared memory?

Hmm, I just found https://www.sqlite.org/uri.html#uriimmutable - seems like rpm may be able to use that when it detects it can't write?

**Comment by pmatilai** [comment]
As per https://www.sqlite.org/wal.html#read_only_databases:

> Even though it is possible to open a read-only WAL-mode database, it is good practice to converted to [PRAGMA journal_mode=DELETE](https://www.sqlite.org/pragma.html#pragma_journal_mode) prior to burning an SQLite database image onto read-only media.

We want to follow good practises so...

Any operation should do a database rebuild to free up any fluff in the files, so I think this should be either a default thing in rebuilddb or an extra flag to it.

**Comment by cgwalters** [comment]
@ffesti the issue I mentioned at Devconf.cz

That said...now that I'm digging in again to reproducible builds, https://github.com/containers/buildah/issues/5592 is higher on my radar, but this is probably in the top 5.

**Comment by cgwalters** [comment]
> so I think this should be either a default thing in rebuilddb 

On this topic, it would be a nontrivial hit to the ergonomics of default container builds if we really start telling people to do:

`RUN dnf -y install foo && dnf clean all && rpm --fix-database`

So probably what should happen is dnf should detect it's in a container and do this by default as part of `dnf clean all` most likely.

**Comment by cgwalters** [comment]
I came across https://github.com/microsoft/rpmoci/blob/47443c4312505222460dfcd185308db7153897bb/src/lockfile/build.rs#L173

**Comment by cgwalters** [comment]
Two things here:

- I finally decided to just copy what rpmoci is doing into rpm-ostree: https://github.com/coreos/rpm-ostree/pull/5244/commits/7b0f07f2f010a0be4bdd26ea518197ad77f1ccab which works around this for bootc images today.
- I also discovered that dnf5 has its *own* sqlite databases nowadays in https://github.com/rpm-software-management/dnf5/issues/2011

**Comment by champtar** [comment]
Found another issue, `rpmtsOrder()` only order the packages to respect the dependencies but it doesn't fully sort the packages list (we are going to sort before `rpmtsOrder()` in rpm-ostree)

**Comment by pmatilai** [comment]
Yeah rpmtsOrder() never promised a stable order if the input changes. I think this is quite intentional on the rpm level to avoid legitimizing ideas like bbb will be installed after aaa, but if somebody wants to sort the packages before inputting to rpm, that's not our concern.

**Comment by cgwalters** [comment]
Bigger picture this stuff I think comes from a conceptual conflict between these package system state files and reproducible image-based systems. It's a *reconcilable* conflict, but basically if I wanted to know "did package X get installed before Y" on an image-based gitops system, I'd just consult the build logs. Is it useful to be able to get that information from the image itself? Sure, I guess, but I wouldn't care at all if it wasn't possible.

Whereas with mutable locally-assembled systems (non-image-based) we end up with a huge proliferation of log files (both `dnf.log` *and* audit logs!), state databases etc. because admins may need to be able to reconstruct who did what when on a per-machine state.

The conflict is: for image based systems we care about bit-for-bit reproducibility a lot (this issue!) whereas for mutable locally-assembled systems one doesn't care about that for the log/state files.

To say this another way, sure I get it that RPM allows multiple installation orders. But why not add e.g. `rpmtsOrderTotal` or something and ensure that everything generating an image does it the same way?

BTW this "installation order" thing of course *shouldn't* matter modulo package bugs, but it'd be interesting to do something like not order alphabetically, but actually use a random seed derived from the package set to permute it each time - at least optionally. That'd really help shake out any such bugs that exist.

**Comment by jlebon** [comment]
Re. package ordering, proposed fixing this in dnf5 for now: https://github.com/rpm-software-management/dnf5/pull/2522.

Agree it'd be nice to have a knob built into librpm though. Perhaps a transaction flag or a global macro (actually, the latter would allow this to be completely transparent to dnf).

**Comment by pmatilai** [comment]
Folks, this ticket is about parking the database. Discussion about package ordering belongs elsewhere.
